### ISSUE TITLE
bug: Memory leak when logger log the traffic with large (about 128K) response body

### ISSUE THREAD

**Issue by caogw279** [apache/apisix#11244]
### Current Behavior

When configured kafka-logger  in global rules with include_resp_body set to true and upstream server response 128K body, the memory will increase during the test and never get down,  if the test continue, the memory usage will continue increase until the OOM happen.

### Expected Behavior

The memory usage will keep in a range during the test and will return to normal after the stress stop.

### Error Logs

No error logs, please have look at output of "free" and the RES of very worker process
<img width="674" alt="apisix_memory_leak" src="https://github.com/apache/apisix/assets/27066781/33642b0c-049b-42f0-bc7d-65dbbed4afcb">

### Steps to Reproduce

1. Run apisix in CentOS 
2. APISIX  Config route with Admin API
3. APISIX  Config kafka-logger to globale_rules with  include_resp_body set to true
4. Upstream server config, make the server return a text/html which  128K or more.
5. Stress the APISIX
6. Monitor the system memory usage (such as via order free -h) 
7. Check the memory usage of every worker process (such as via top)

### Environment

- APISIX version (run `apisix version`): apisix 3.8.0
- Operating system (run `uname -a`): Linux localhost.localdomain 3.10.0-1160.105.1.el7.x86_64 #1 SMP Thu Dec 7 15:39:45 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): openresty/1.21.4.2
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`): 3.5.0
- APISIX Dashboard version, if relevant: No APISIX Dashboard used.
- Plugin runner version, for issues related to plugin runners: no external plugin is used
- LuaRocks version, for installation issues (run `luarocks --version`): No luarocks is installed.

**Comment by caogw279** [comment]
Addition:  I test it with wrk as client and take the nginx as the upstream server, and change it's index.html with 512k bytes characters,  can't reproduce the issue, resident memory of every openresty process will keep at 1 - 1.5 G  no matter how long wrk stress it.

**Comment by caogw279** [comment]
From the document [,](https://apisix.apache.org/docs/apisix/control-api/) I found a method which will Triggers a full garbage collection in the HTTP subsystem, it's works,  the memory not leak ? just hold by GC ? but why doesn't GC collection work by itself, why the memory is exhausted and GC still not work  ?

**Comment by shreemaan-abhishek** [comment]
@caogw279, please share the details on how you stress test APISIX.

**Comment by shreemaan-abhishek** [comment]
@Baoyuantop, can you try load testing with this patch: https://github.com/apache/apisix/pull/12035/files

last time I can confirm I was able to fix the memory leak bug by this patch. But could not document and prove it as I did not have proper wrk tool setup in my development machine.

**Comment by AlinsRan** [comment]
I reproduced this in containers (APISIX 3.8.0 and dev/3.17.0) and traced it through the LuaJIT internals. The growth is **not** in `hold_body_chunk`. Here is what actually causes it and how to bound it.

## TL;DR

- Not a leak — a forced GC (`/v1/gc`) reclaims most of it.
- Cause: an **unbounded logger backlog** of large response bodies, kept resident by **LuaJIT's GC + allocator behavior**.
- Fix: bound the backlog — already possible today with `max_pending_entries` or `max_resp_body_bytes`.
- 3.8.0 and current dev behave identically.

## Where the memory is

`hold_body_chunk` is bounded per request (≤ `max_resp_body_bytes`) and freed when the request ends. Proxying the same large bodies *without* a logger stays flat (~39 MB), so it is not the source.

The memory lives in the **batch-processor backlog**:

- With `include_resp_body`, every queued log entry holds a full response body.
- Each flushed batch is shipped by its own timer, and **nothing caps the total bytes or the number of in-flight batches**.
- When the sink (Kafka/HTTP) is slower than the request rate, in-flight batches pile up as `rate × sink_latency × body_size`.

These entries are still referenced (waiting to be sent), so the GC correctly cannot collect them. This is the real, fixable cause.

## Why it doesn't drop after the load stops

Two LuaJIT properties (verified against `lj_alloc.c`):

**1. The GC runs only on allocation.** It advances its mark-and-sweep in steps triggered by new allocations. When traffic stops there are no allocations, so the dead entries aren't swept until the next request or an explicit `collectgarbage()`. That is why RSS stays at the peak while idle, and why `/v1/gc` frees it.

**2. The GC never moves objects, and the allocator only partially returns memory.** LuaJIT allocates via `lj_alloc` (a dlmalloc):

- blocks ≥ 128 KB → `mmap` / `munmap`, returned to the OS on free;
- smaller blocks → arena free-lists, released only by trimming the top of the heap (≥ 2 MB).

Bodies of 64–512 KB straddle the 128 KB line: the large ones are returned on free, the small ones leave holes the non-moving GC can't compact.

So after a GC the Lua objects are gone, yet RSS sits at the fragmented high-water. In my test `/v1/gc` reclaimed ~952 → ~193 MB — most returns, the fragmented remainder stays — which is also why it helps more in some environments than others.

## Measured

http-logger + `include_resp_body`, bodies 64–512 KB, slow sink, ~400 rps:

| config | under load | idle after |
|---|---|---|
| default (unbounded) | ~1.3 GB | ~850 MB (3.8.0 882 / dev 842), no drop |
| `max_pending_entries=500` | ~51 MB | ~51 MB |
| `max_resp_body_bytes=16K` | ~115 MB | ~92 MB |

## Fix

Bound the backlog so the live set never grows large enough to fragment the allocator. No code change needed:

- **`max_pending_entries`** — caps the entry count; keeps full bodies, drops the excess.
- **`max_resp_body_bytes`** — caps per-entry body size (truncates).

Trade-off worth stating: bounded memory + full bodies + a sink that can't keep up ⇒ excess must be dropped. No buffer setting is both bounded and lossless under sustained overload.

`max_buffer_bytes` (proposed in #13581) is an optional byte-precise variant — not required, since `max_pending_entries` already covers this case.

**Comment by AlinsRan** [comment]
Closing this out with the conclusion from the analysis above.

Summary: this is not a leak in `hold_body_chunk` — it is an unbounded batch-processor backlog of large response bodies, made visible by LuaJIT's non-moving GC and the `lj_alloc` allocator (which only returns blocks >= 128 KB to the OS and trims the arena at >= 2 MB), so RSS stays at the fragmented high-water mark after the load stops.

Mitigations, both available today, no code change needed:

- `max_pending_entries` (plugin metadata, e.g. `/apisix/admin/plugin_metadata/kafka-logger`) — caps the backlog and drops the excess. Added in #12338, so it is available since **3.13.0**; it did not exist in 3.8.0, which is why there was no way out at the time this was reported.
- `max_resp_body_bytes` (plugin attribute, default 512 KB) — caps the body kept per entry.

Trade-off, stated plainly: bounded memory + full bodies + a sink that can't keep up ⇒ excess must be dropped. No buffer setting is both bounded and lossless under sustained overload.

If you have any doubts, please reopen this PR.
